File: 73.0-1KnapsackAlgo.py
# The 0/1 Knapsack problem is a classic dynamic programming problem 
# where we are given a set of items, each with a weight and value, and a 
# knapsack with a maximum weight capacity. The goal is to maximize the total 
# value of items placed in the knapsack without exceeding its weight. 
# Each item can be either taken or left (0 or 1).

# Explanation

# We have n items, each with a value[i] and weight[i].

# The knapsack can carry a maximum weight W.

# The key idea is to decide for each item: include it or exclude it.

# We use dynamic programming to avoid recalculating the same subproblems.

# We define a 2D DP table dp[i][w] where:

# i represents the first i items

# w represents the current weight capacity

# dp[i][w] stores the maximum value achievable with i items and weight w.

# The recurrence relation:

# If weight of current item weight[i-1] <= w:
# dp[i][w] = max(dp[i-1][w], value[i-1] + dp[i-1][w - weight[i-1]])

# Otherwise:
# dp[i][w] = dp[i-1][w]

values = [60, 100, 120]
weights = [10, 20, 30]

W = 50

n = len(values)

dp = [[0 for _ in range(W + 1)] for _ in range(n + 1)]

for i in range(1, n + 1):
    for w in range(1, W + 1):
        if weights[i - 1] <= w:
            dp[i][w] = max(dp[i - 1][w], values[i - 1] + dp[i - 1][w - weights[i - 1]])
        else:
            dp[i][w] = dp[i - 1][w]

print("Maximum value in the Knapsack =", dp[n][W])

File: 73.0-1KnapsackAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/* -----------------------------------------------------------------------  */

public class Knapsack {

    public static void main(String[] args) {

        // Values (profit) of items
        int[] values = {60, 100, 120};
        // Corresponding weights of items
        int[] weights = {10, 20, 30};
        // Maximum weight capacity of knapsack
        int W = 50;
        // Number of items
        int n = values.length;

        // Initialize DP table (n+1) x (W+1) with 0
        int[][] dp = new int[n+1][W+1];

        // Build the DP table in bottom-up manner
        for (int i = 1; i <= n; i++) { // Loop through items
            for (int w = 1; w <= W; w++) { // Loop through all capacities
                if (weights[i-1] <= w) { // If current item's weight <= current capacity
                    // Maximum of including or excluding the item
                    dp[i][w] = Math.max(dp[i-1][w], values[i-1] + dp[i-1][w - weights[i-1]]);
                } else {
                    // Cannot include the item, take value from previous row
                    dp[i][w] = dp[i-1][w];
                }
            }
        }

        // Maximum value achievable with given weight
        System.out.println("Maximum value in knapsack = " + dp[n][W]);
    }
}
